### PR TITLE
Feat/prod3 12

### PR DESCRIPTION
**Pull request by mendesmcmg**
# Description

- Adds google sign in for users

## Does this pull request close any open issues?

- [#prod3-12](https://ribon.atlassian.net/browse/PROD3-12?atlOrigin=eyJpIjoiNDVhYmRmNTQzNjBjNGY0NmJiZmI2MzNhMGIwYTAxNzgiLCJwIjoiaiJ9)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Dangerous changes
- [ ] Chore (change that does not effect how the code works, eg: change color)
- [ ] Tests

## How to test

- go to hoppscotch and use the google auth request in Auth - Dapp
- add this to the request body:
 ```
{
  "data": {
    "id_token": "your_actual_id_token_here"
  }
}

```

### CODE REVIEW

File: app/models/account.rb [hunk 1]
```diff
@@ -21,13 +21,28 @@
...
-
   delegate :email, to: :user
+  delegate :unconfirmed_email, to: :user
+
+  def self.create_user_for_google(data)
+    user = User.find_or_create_by(email: data['email'])
+    account = find_or_initialize_by(user:)
+    account.assign_attributes(
+      provider: 'google_oauth2',
+      uid: data['email'],
```
Comment (mendesmcmg) [inline app/models/account.rb:41]: me perguntando se deveríamos encriptar isso

File: app/models/users/user.rb [hunk 2]
```diff
@@ -78,6 +78,10 @@ def donate_app
     false
   end
 
+  def unconfirmed_email
```
Comment (mendesmcmg) [inline app/models/users/user.rb:81]: fixz pra dar um fix num erro do devise que pede essa propriedade, mas aceito sugestões melhores

File: app/models/account.rb (hunk 1 above)
Comment (NickNish09) [reply app/models/account.rb:41]: acredito que fica ok na verdade o uid ser o email, o proprio devise pra autenticação padrão usa

File: app/commands/auth/accounts/set_account_tokens.rb [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+        @id_token = id_token
+      end
+
+      def call
+        with_exception_handle do
+          response = Request::ApiRequest.get(google_api_url)
+
+          @account = Account.create_user_for_google(response)
+          access_token, refresh_token = Jwt::Auth::Issuer.call(@account)
+          @account.save
```
Comment (NickNish09) [inline app/commands/auth/accounts/set_account_tokens.rb:20]: maybe we should use `@account.save!` before the token creation? To ensure that the account is valid and created. If not the command will return an error. What do you think

File: config/routes.rb [hunk 4]
```diff
@@ -237,7 +237,9 @@
     
   namespace :users do
     namespace :v1 do
+      mount_devise_token_auth_for 'Account', at: 'auth', skip: [:omniauth_callbacks]
```
Comment (NickNish09) [inline config/routes.rb:240]: are we using the devise routes as well? I don't think we need them since we have the google route already

**Comment by NickNish09** [comment]
I don't know if there is some test codes to the `app/commands/auth/accounts/set_account_tokens.rb` file, but I think would be nice to put some specs for it

**Review by leodhb — APPROVED** [review]
LGTM

**Pull request merged**